snapcraft/internal/build_providers/_snap.py

```diff
@@ -0,0 +1,347 @@
+        elif is_installed and self._latest_revision != host_snap_info["revision"]:
+            op = _SnapOp.INJECT
+        else:
+            # This is a programmatic error
+            raise RuntimeError(
+                "Unhandled scenario for {!r} (host installed: {}, latest_revision {})".format(
+                    self.snap_name, is_installed, self._latest_revision
+                )
+            )
+
+        self.__required_operation = op
+        return op
+
+    def get_assertions(self) -> List[List[str]]:
+        op = self.get_op()
+        if op != op.INJECT:
+            return []
+        host_snap_repo = self._get_snap_repo()
+        host_snap_info = host_snap_repo.get_local_snap_info()
+
+        if host_snap_info["revision"].startswith("x"):
+            return []
+
+        assertions = []  # type: List[List[str]]
+        assertions.append(
+            ["snap-declaration", "snap-name={}".format(host_snap_repo.name)]
+        )
+        assertions.append(
+            [
+                "snap-revision",
+                "snap-revision={}".format(host_snap_info["revision"]),
+                "snap-id={}".format(host_snap_info["id"]),
+            ]
+        )
+        return assertions
+
+    def _set_data(self) -> None:
+        op = self.get_op()
+        host_snap_repo = self._get_snap_repo()
+        install_cmd = ["sudo", "snap"]
+
+        if op == _SnapOp.INJECT:
+            install_cmd.append("install")
+            host_snap_info = host_snap_repo.get_local_snap_info()
+            snap_revision = host_snap_info["revision"]
+
+            if snap_revision.startswith("x"):
+                install_cmd.append("--dangerous")
+
+            if host_snap_info["confinement"] == "classic":
+                install_cmd.append("--classic")
+
+            snap_file_name = "{}_{}.snap".format(host_snap_repo.name, snap_revision)
+            install_cmd.append(os.path.join(self._snap_dir, snap_file_name))
+        elif op == _SnapOp.INSTALL or op == _SnapOp.REFRESH:
+            install_cmd.append(op.name.lower())
+            store_snap_info = storeapi.StoreClient().cpi.get_package(
+                self.snap_name, "latest/stable", self._snap_arch
+            )
+            snap_revision = store_snap_info["revision"]
+            confinement = store_snap_info["confinement"]
+            if confinement == "classic":
+                install_cmd.append("--classic")
+            install_cmd.append(host_snap_repo.name)
+        elif op == _SnapOp.NOP:
+            install_cmd = []
+            snap_revision = None
+
+        self.__install_cmd = install_cmd
+        self.__revision = snap_revision
+
+    def get_revision(self) -> str:
+        if self.__revision is None:
+            self._set_data()
+        return self.__revision
+
+    def get_install_cmd(self) -> List[str]:
+        if self.__install_cmd is None:
+            self._set_data()
+        return self.__install_cmd
+
+
+_STORE_ASSERTION = [
```

Comment:
Can we move this up to the top of the file, please?

Reply:
sure